python/tests/ibd.py

```diff
@@ -0,0 +1,421 @@
+        return (
+            self.left == other.left
+            and self.right == other.right
+            and self.node == other.node
+        )
+
+    def __lt__(self, other):
+        return (self.node, self.left, self.right) < (
+            other.node,
+            other.left,
+            other.right,
+        )
+
+
+class SegmentList:
+    """
+    A class representing a list of segments that are descended from a given ancestral
+    node via a particular child of the ancestor.
+    Each SegmentList keeps track of the first and last segment in the list, head and
+    tail. The next attribute points to another SegmentList, allowing SegmentList
+    objects to be 'chained' to one another.
+    """
+
+    def __init__(self, head=None, tail=None, next_list=None):
+        self.head = head
+        self.tail = tail
+        self.next = next_list
+
+    def __str__(self):
+        s = "head={},tail={},next={}".format(self.head, self.tail, repr(self.next))
+        return s
+
+    def __repr__(self):
+        if self.head is None:
+            s = "[{}]".format(repr(None))
+        elif self.head == self.tail:
+            s = "[{}]".format(repr(self.head))
+        elif self.head.next == self.tail:
+            s = "[{}, {}]".format(repr(self.head), repr(self.tail))
+        else:
+            s = "[{}, ..., {}]".format(repr(self.head), repr(self.tail))
+        return s
+
+    def __len__(self):
+        # Returns the number of segments in the list.
+        count = 0
+        seg = self.head
+        while seg is not None:
+            count += 1
+            seg = seg.next
+        return count
+
+    def add(self, other):
+        """
+        Use to append another SegmentList, or a single segment.
+        SegmentList1.add(SegmentList2) will modify SegmentList1 so that
+        SegmentList1.tail.next = SegmentList2.head
+        SegmentList1.add(Segment1) will add Segment1 to the tail of SegmentList1
+        """
+        assert isinstance(other, SegmentList) or isinstance(other, Segment)
+
+        if isinstance(other, SegmentList):
+            if self.head is None:
+                self.head = other.head
+                self.tail = other.tail
+            else:
+                self.tail.next = other.head
+                self.tail = other.tail
+        elif isinstance(other, Segment):
+            if self.head is None:
+                self.head = other
+                self.tail = other
+            else:
+                self.tail.next = other
+                self.tail = other
+
+
+class IbdFinder:
+    """
+    Finds all IBD relationships between specified samples in a tree sequence.
+    """
+
+    def __init__(self, ts, samples=None, min_length=0, max_time=None):
+
+        self.ts = ts
+        # Note: samples *must* be in order of ascending node ID
+        if samples is None:
+            self.samples = ts.samples()
+        else:
+            self.samples = samples
+        self.samples.sort()
```

Comment:
I don't think we should do this. The ordering of the samples is defined by the user and it would be unexpected to change this. Instead, we do something like
```python
self.sample_index_map = np.zeros(ts.num_nodes, dtype=int) - 1
for index, u in enumerate(self.samples):
    self.sample_index_map[u] = index
```
Then, when we want to know if something is a sample we see if ``self.sample_index_map[u] != tskit.NULL``.